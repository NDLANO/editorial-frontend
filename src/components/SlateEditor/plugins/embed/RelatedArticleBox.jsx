/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { injectT } from 'ndla-i18n';
import { connect } from 'react-redux';
import Types from 'slate-prop-types';
import { RelatedArticleList, RelatedArticle } from 'ndla-ui';
import { toggleRelatedArticles } from 'ndla-article-scripts';
import get from 'lodash/fp/get';
import { searchArticles } from '../../../../modules/article/articleApi';
import { fetchArticleResource } from '../../../../modules/taxonomy/taxonomyApi';
import { getLocale } from '../../../../modules/locale/locale';
import { EditorShape } from '../../../../shapes';
import { mapping } from '../utils/relatedArticleMapping';
import EditRelated from './EditRelated';

const nodeProps = ids => ({
  data: {
    resource: 'related-content',
    'article-ids': ids,
  },
});

class RelatedArticleBox extends React.Component {
  constructor() {
    super();
    this.state = { items: [], editMode: true };
    this.removeArticle = this.removeArticle.bind(this);
    this.fetchRelated = this.fetchRelated.bind(this);
    this.onInsertBlock = this.onInsertBlock.bind(this);
    this.updateNodeAttributes = this.updateNodeAttributes.bind(this);
  }

  componentDidMount() {
    const { embed } = this.props;
    if (embed['article-ids']) {
      embed['article-ids'].split(',').map(it => this.fetchRelated(it));
    }
  }

  componentDidUpdate() {
    // need to re-add eventhandler on button
    if (!this.state.editMode && this.state.items.length > 2) {
      toggleRelatedArticles();
    }
  }

  onInsertBlock(newArticle) {
    if (!this.state.items.find(it => it.id === parseInt(newArticle, 10))) {
      // get resource and add to state
      this.fetchRelated(newArticle);

      // update slate block attributes
      const currentIds = this.state.items.map(it => it.id).join(',');
      this.updateNodeAttributes(
        `${currentIds}${currentIds ? ',' : ''}${newArticle}`,
      );
    }
  }

  updateNodeAttributes(ids) {
    const { editor, node } = this.props;
    editor.change(change => change.setNodeByKey(node.key, nodeProps(ids)));
  }

  async fetchRelated(it) {
    const { locale } = this.props;

    const [article, resource] = await Promise.all([
      searchArticles(`${it}`, locale),
      fetchArticleResource(it, locale),
    ]);

    this.setState(prevState => ({
      items: [...prevState.items, { ...article, resource }],
      editMode: false,
    }));
  }

  removeArticle(i, e) {
    e.stopPropagation();

    const newItems = this.state.items.filter((_, ind) => i !== ind);
    this.setState({ items: newItems });

    // remove from slate attribute
    this.updateNodeAttributes(newItems.map(it => it.id).join(','));
  }

  render() {
    const { attributes, onRemoveClick, locale, t } = this.props;
    const { editMode, items } = this.state;

    const resourceType = item =>
      item.resourceTypes
        ? item.resourceTypes.find(it => mapping(it.id))
        : { id: '' };

    return this.state.editMode ? (
      <EditRelated
        {...{
          onRemoveClick,
          removeArticle: this.removeArticle,
          resourceType,
          items,
          editMode,
          locale,
          onInsertBlock: this.onInsertBlock,
          onExit: () => this.setState({ editMode: false }),
        }}
      />
    ) : (
      <div
        role="button"
        tabIndex={0}
        onClick={e => {
          e.stopPropagation();
          this.setState({ editMode: true });
        }}
        {...attributes}>
        <RelatedArticleList
          messages={{
            title: t('form.related.title'),
            showMore: t('form.related.showMore'),
            showLess: t('form.related.showLess'),
          }}>
          {items.map(
            item =>
              !item.id ? (
                'Invalid article'
              ) : (
                <RelatedArticle
                  {...mapping(resourceType(item).id)}
                  title={get('title.title', item)}
                  key={item.id}
                  introduction={get('metaDescription.metaDescription', item)}
                  to={`/learning-resource/${item.id}/edit/${locale}`}
                />
              ),
          )}
        </RelatedArticleList>
      </div>
    );
  }
}

RelatedArticleBox.propTypes = {
  attributes: PropTypes.shape({
    'data-key': PropTypes.string.isRequired,
  }),
  editor: EditorShape,
  node: Types.node.isRequired,
  locale: PropTypes.string.isRequired,
  onRemoveClick: PropTypes.func,
  embed: PropTypes.shape({
    resource: PropTypes.string,
    'article-ids': PropTypes.string,
    relatedArticle: PropTypes.string,
  }),
};

const mapStateToProps = state => ({
  locale: getLocale(state),
});

export default injectT(connect(mapStateToProps)(RelatedArticleBox));
