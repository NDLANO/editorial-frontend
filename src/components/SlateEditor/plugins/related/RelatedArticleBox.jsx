/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { Transforms } from 'slate';
import { ReactEditor } from 'slate-react';
import PropTypes from 'prop-types';
import { uuid } from '@ndla/util';
import { injectT } from '@ndla/i18n';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { css } from '@emotion/core';
import { RelatedArticleList } from '@ndla/ui';
import { toggleRelatedArticles } from '@ndla/article-scripts';
import { convertFieldWithFallback } from '../../../../util/convertFieldWithFallback';
import { fetchDraft } from '../../../../modules/draft/draftApi';
import { queryResources } from '../../../../modules/taxonomy';
import { getLocale } from '../../../../modules/locale/locale';
import { EditorShape, AttributesShape } from '../../../../shapes';
import EditRelated from './EditRelated';
import handleError from '../../../../util/handleError';
import RelatedArticle from './RelatedArticle';
import { ARTICLE_EXTERNAL } from '../../../../constants';

const mapRelatedArticle = (article, resource) => ({
  ...article,
  resource,
  id: `${article.id}`,
  title: convertFieldWithFallback(article, 'title', article.title),
});

export class RelatedArticleBox extends React.Component {
  constructor() {
    super();
    this.state = { articles: [], editMode: false };
    this.fetchArticle = this.fetchArticle.bind(this);
    this.updateArticles = this.updateArticles.bind(this);
    this.insertExternal = this.insertExternal.bind(this);
    this.onInsertBlock = this.onInsertBlock.bind(this);
    this.openEditMode = this.openEditMode.bind(this);
    this.setNodeData = this.setNodeData.bind(this);
  }

  componentDidMount() {
    const {
      element: { data },
    } = this.props;
    if (data && data.nodes) {
      const articleNodes = data.nodes;
      this.fetchArticles(articleNodes).then(articles =>
        this.setState({ articles: articles.filter(a => !!a) }),
      );
    } else {
      this.setState({ editMode: true });
    }
  }

  componentDidUpdate() {
    // need to re-add eventhandler on button
    if (!this.state.editMode && this.state.articles.length > 2) {
      toggleRelatedArticles();
    }
  }

  onInsertBlock(newArticle) {
    if (!this.state.articles.find(it => it.id === parseInt(newArticle, 10))) {
      // get resource and add to state
      this.fetchArticle(newArticle).then(article => {
        if (article) {
          this.setState(oldState => ({
            articles: [...oldState.articles, article],
          }));
          this.setNodeData();
        }
      });
    }
  }

  setNodeData() {
    const { editor, element } = this.props;
    const { articles } = this.state;
    const path = ReactEditor.findPath(editor, element);
    Transforms.setNodes(
      editor,
      {
        data: {
          nodes: articles.map(article =>
            article.id === ARTICLE_EXTERNAL
              ? {
                  resource: 'related-content',
                  url: article.url,
                  title: article.title,
                }
              : { resource: 'related-content', 'article-id': article.id },
          ),
        },
      },
      { at: path, voids: true },
    );
  }

  structureExternal(url, title) {
    return {
      id: ARTICLE_EXTERNAL,
      tempId: uuid(),
      url,
      title,
      description: '',
    };
  }

  async fetchArticle(id) {
    try {
      const { locale } = this.props;
      const [article, resource] = await Promise.all([
        fetchDraft(id, locale),
        queryResources(id, locale),
      ]);
      if (article) {
        return mapRelatedArticle(article, resource);
      }
    } catch (error) {
      handleError(error);
    }
  }
  async fetchArticles(nodes) {
    return Promise.all(
      nodes.map(node => {
        const articleId = node['article-id'];
        if (articleId) {
          return this.fetchArticle(articleId);
        } else if (node.title) {
          return this.structureExternal(node.url, node.title);
        }
        return undefined;
      }),
    );
  }

  async insertExternal(url, title) {
    // await get description meta data
    this.setState(
      prevState => ({
        articles: [...prevState.articles, this.structureExternal(url, title)],
        editMode: false,
      }),
      this.setNodeData,
    );
  }

  updateArticles(newArticles) {
    this.setState({ articles: newArticles.filter(a => !!a) }, this.setNodeData);
  }

  openEditMode(e) {
    e.stopPropagation();
    this.setState({ editMode: true });
  }

  render() {
    const { attributes, onRemoveClick, locale, t, children } = this.props;
    const { editMode, articles } = this.state;

    if (editMode) {
      return (
        <EditRelated
          onRemoveClick={onRemoveClick}
          articles={articles}
          locale={locale}
          insertExternal={this.insertExternal}
          onInsertBlock={this.onInsertBlock}
          onExit={() => this.setState({ editMode: false })}
          updateArticles={this.updateArticles}
          {...attributes}
        />
      );
    }

    return (
      <div
        role="button"
        draggable
        contentEditable={false}
        tabIndex={0}
        data-testid="relatedWrapper"
        onClick={this.openEditMode}
        onKeyPress={this.openEditMode}
        css={css`
          & article > p {
            font-family: Source Sans Pro !important;
          }
        `}
        {...attributes}>
        <RelatedArticleList
          messages={{
            title: t('form.related.title'),
            showMore: t('form.related.showMore'),
            showLess: t('form.related.showLess'),
          }}>
          {articles.map((item, i) =>
            !item.id ? (
              t('form.content.relatedArticle.invalidArticle')
            ) : (
              <RelatedArticle key={uuid()} numberInList={i} item={item} />
            ),
          )}
        </RelatedArticleList>
        {children}
      </div>
    );
  }
}

RelatedArticleBox.propTypes = {
  attributes: AttributesShape,
  editor: EditorShape.isRequired,
  element: PropTypes.any,
  locale: PropTypes.string.isRequired,
  onRemoveClick: PropTypes.func,
};

const mapStateToProps = state => ({
  locale: getLocale(state),
});

export default compose(injectT, connect(mapStateToProps, null))(RelatedArticleBox);
