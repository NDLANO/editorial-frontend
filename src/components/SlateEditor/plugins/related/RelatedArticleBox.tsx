/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { useEffect, useRef, useState, MouseEvent, ReactNode } from 'react';
import { Editor, Transforms } from 'slate';
import { ReactEditor, RenderElementProps } from 'slate-react';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import { RelatedContentEmbedData, RelatedContentMetaData } from '@ndla/types-embed';
import { RelatedArticleListV2, RelatedContentEmbed } from '@ndla/ui';
import { IconButtonV2 } from '@ndla/button';
import { Pencil, TrashCanOutline } from '@ndla/icons/action';
import { fetchDraft } from '../../../../modules/draft/draftApi';
import EditRelated from './EditRelated';
import { RelatedElement } from '.';
import { useTaxonomyVersion } from '../../../../containers/StructureVersion/TaxonomyVersionProvider';
import { fetchNodes } from '../../../../modules/nodes/nodeApi';

interface Props {
  attributes: RenderElementProps['attributes'];
  editor: Editor;
  element: RelatedElement;
  onRemoveClick: (e: MouseEvent<HTMLButtonElement>) => void;
  children: ReactNode;
}

const externalEmbedToMeta = async (
  embedData: RelatedContentEmbedData,
  seq: number,
): Promise<RelatedContentMetaData> => {
  return {
    resource: 'related-content',
    embedData,
    seq,
    status: 'success',
    data: undefined,
  };
};

const internalEmbedToMeta = async (
  embedData: RelatedContentEmbedData,
  seq: number,
  language: string,
  taxonomyVersion: string,
): Promise<RelatedContentMetaData> => {
  const article = await fetchDraft(embedData.articleId!, language).catch(() => undefined);
  const nodes = await fetchNodes({
    taxonomyVersion,
    contentURI: `urn:article:${embedData.articleId}`,
    language: language,
  }).catch(() => undefined);

  if (!!article && !!nodes?.length) {
    return {
      resource: 'related-content',
      seq,
      embedData,
      status: 'success',
      data: {
        article,
        resource: nodes[0]!,
      },
    };
  } else {
    return {
      resource: 'related-content',
      seq,
      embedData,
      status: 'error',
      message: 'Failed to fetch data',
    };
  }
};

const embedsToMeta = async (
  embeds: RelatedContentEmbedData[],
  language: string,
  taxonomyVersion: string,
) => {
  const promises = embeds.map((embed, i) => {
    if (embed.articleId) {
      return internalEmbedToMeta(embed, i, language, taxonomyVersion);
    } else {
      return externalEmbedToMeta(embed, i);
    }
  });
  return await Promise.all(promises);
};

const StyledIconButton = styled(IconButtonV2)`
  width: 40px;
  height: 40px;
`;

const RelatedArticleBox = ({ attributes, editor, element, onRemoveClick, children }: Props) => {
  const { t, i18n } = useTranslation();
  const { taxonomyVersion } = useTaxonomyVersion();
  const [editMode, setEditMode] = useState(false);
  const [embeds, setEmbeds] = useState<RelatedContentMetaData[]>([]);
  const firstRender = useRef(true);

  useEffect(() => {
    if (!firstRender.current) {
      return;
    }
    if (!!element?.data?.length && firstRender.current) {
      embedsToMeta(element.data, i18n.language, taxonomyVersion).then(setEmbeds);
      firstRender.current = false;
    } else {
      setEditMode(true);
    }
  }, [element.data, i18n.language, taxonomyVersion]);

  const insertExternal = async (title: string, url: string) => {
    const newEmbed: RelatedContentEmbedData = {
      resource: 'related-content',
      title,
      url,
    };
    const nodeData = (element.data ?? []).concat(newEmbed);
    setNodeData(nodeData);
    const newMetaData = await externalEmbedToMeta(newEmbed, embeds.length + 1);
    setEmbeds((embeds) => embeds.concat(newMetaData));
  };

  const insertInternal = async (articleId: string) => {
    const existingNodes = element.data;
    const exists = existingNodes.some((embed) => embed.articleId === articleId);
    if (exists) {
      return;
    }
    const newEmbed: RelatedContentEmbedData = { resource: 'related-content', articleId };
    const embed = await internalEmbedToMeta(
      { resource: 'related-content', articleId },
      existingNodes.length + 1,
      i18n.language,
      taxonomyVersion,
    );
    setEmbeds((embeds) => embeds.concat(embed));
    setNodeData(existingNodes.concat(newEmbed));
  };

  const setNodeData = (data: RelatedContentEmbedData[]) => {
    const path = ReactEditor.findPath(editor, element);
    Transforms.setNodes<RelatedElement>(editor, { data }, { at: path, voids: true });
  };

  const deleteElement = () => {
    const path = ReactEditor.findPath(editor, element);
    ReactEditor.focus(editor);
    Transforms.select(editor, path);
    Transforms.removeNodes(editor, { at: path });
  };

  const updateArticles = (newEmbeds: RelatedContentMetaData[]) => {
    setEmbeds(newEmbeds.filter((a) => !!a));
    setNodeData(newEmbeds.map((embed) => embed.embedData));
  };

  return (
    <div contentEditable={false} {...attributes}>
      {editMode && (
        <EditRelated
          data-testid="editRelated"
          onRemoveClick={onRemoveClick}
          embeds={embeds}
          insertExternal={insertExternal}
          onInsertBlock={insertInternal}
          onExit={() => setEditMode(false)}
          updateArticles={updateArticles}
        />
      )}
      <RelatedArticleListV2
        data-testid="relatedWrapper"
        headingButtons={
          <div>
            <StyledIconButton
              onClick={() => setEditMode(true)}
              aria-label={t('form.edit')}
              variant="ghost"
            >
              <Pencil />
            </StyledIconButton>
            <StyledIconButton
              onClick={deleteElement}
              aria-label={t('delete')}
              variant="ghost"
              colorTheme="danger"
            >
              <TrashCanOutline />
            </StyledIconButton>
          </div>
        }
      >
        {embeds.map((embed) => (
          <RelatedContentEmbed key={`related-${embed.seq}`} embed={embed} />
        ))}
      </RelatedArticleListV2>
      {children}
    </div>
  );
};

export default RelatedArticleBox;
