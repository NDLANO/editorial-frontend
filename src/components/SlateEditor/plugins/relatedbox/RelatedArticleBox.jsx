/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import BEMHelper from 'react-bem-helper';
import Types from 'slate-prop-types';
import {
  Button,
  RelatedArticleList,
  RelatedArticle,
  ContentTypeBadge,
  constants,
} from 'ndla-ui';
import { Search } from 'ndla-icons/common';
import { get } from 'lodash';
import { Cross } from 'ndla-icons/action';
import Lightbox from '../../../Lightbox';
import { Portal } from '../../../../components/Portal';
import { EditorShape } from '../../../../shapes';
import { searchArticles } from '../../../../modules/search/searchApi';
import AsyncDropdown from '../../../../components/Dropdown/asyncDropdown/AsyncDropdown';

const classes = new BEMHelper({
  name: 'relatedBox',
  prefix: 'c-',
});

class RelatedArticleBox extends React.Component {
  constructor() {
    super();
    this.state = { items: [], isOpened: true };
    this.onRemoveClick = this.onRemoveClick.bind(this);
    this.onChange = this.onChange.bind(this);
    this.searchArticles = this.searchArticles.bind(this);
    this.toggleOpen = this.toggleOpen.bind(this);
    this.closeAndReturnFocusToEditor = this.closeAndReturnFocusToEditor.bind(
      this,
    );
  }

  onRemoveClick() {
    const { editor, node } = this.props;
    const next = editor.value.change().removeNodeByKey(node.key);
    editor.onChange(next);
  }
  onChange(selected) {
    console.log(selected);
    this.state.items.push(selected);
    this.toggleOpen();
  }

  async searchArticles(input) {
    const query = `?type=articles&query=${input}`;
    const response = await searchArticles(query, 'nb');
    return response.results;
  }

  toggleOpen() {
    console.log(this.state.isOpened);
    this.setState(prevState => ({
      isOpened: !prevState.isOpened,
    }));
  }

  closeAndReturnFocusToEditor() {
    const { editor } = this.props;
    editor.value.change().focus();
    this.toggleOpen();
  }

  render() {
    const { children, attributes } = this.props;
    return (
      <div {...attributes} {...classes()}>
        <Portal isOpened={this.state.isOpened}>
          <Lightbox display big onClose={this.closeAndReturnFocusToEditor}>
            <AsyncDropdown
              valueField="id"
              name="relatedArticleSearch"
              selectedItem={this.state.id}
              textField="title.title"
              placeholder={'placeholder'}
              label={'label'}
              apiAction={this.searchArticles}
              messages={{
                emptyFilter: 'empty',
                emptyList: 'empty list',
              }}
              onChange={this.onChange}
            />
            {this.props.children}
          </Lightbox>
        </Portal>
        <RelatedArticleList messages={{ title: 'Relaterte arikler' }}>
          {this.state.items.map(item => (
            <RelatedArticle
              key={item.id}
              icon={
                <ContentTypeBadge
                  background
                  type={constants.contentTypes.SUBJECT_MATERIAL}
                />
              }
              title={item.title.title}
              introduction={get(item, 'introduction.introduction')}
              to={`/article/${item.id}`}
              modifier="subject-material"
            />
          ))}
        </RelatedArticleList>
        <Button
          stripped
          onClick={this.toggleOpen}
          {...classes('add-related-button')}>
          <Search />
        </Button>
        <Button
          stripped
          onClick={this.onRemoveClick}
          {...classes('delete-button')}>
          <Cross />
        </Button>
      </div>
    );
  }
}

RelatedArticleBox.propTypes = {
  attributes: PropTypes.shape({
    'data-key': PropTypes.string.isRequired,
  }),
  node: Types.node.isRequired,
  editor: EditorShape,
};

export default RelatedArticleBox;
