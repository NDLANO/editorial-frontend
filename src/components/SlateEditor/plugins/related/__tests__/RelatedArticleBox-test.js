/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import nock from 'nock';
import { render, Simulate, wait } from 'react-testing-library';
import { RelatedArticleBox } from '../RelatedArticleBox';
import IntlWrapper from '../../../../../util/__tests__/IntlWrapper';

const wrapper = () =>
  render(
    <IntlWrapper>
      <RelatedArticleBox
        t={() => 'injected'}
        editor={{ onChange: () => {}, change: () => {} }}
        locale="nb"
        node={{}}
      />
    </IntlWrapper>,
  );

test('it goes in and out of edit mode', async () => {
  nock('http://ndla-api')
    .get(
      '/article-api/v2/articles/?language=undefined&fallback=true&type=articles&query=',
    )
    .reply(200, {});
  const { getByTestId, container } = wrapper();

  Simulate.click(getByTestId('relatedWrapper'));
  Simulate.click(getByTestId('showAddExternal'));
  expect(container.firstChild).toMatchSnapshot();

  const input = getByTestId('addExternalUrlInput');
  const inputTitle = getByTestId('addExternalTitleInput');
  input.value = 'www.vg.no';
  inputTitle.value = 'Title';
  Simulate.change(input);
  Simulate.change(inputTitle);

  Simulate.click(getByTestId('taxonomyLightboxButton'));
  await wait();

  expect(container.firstChild).toMatchSnapshot();
});
