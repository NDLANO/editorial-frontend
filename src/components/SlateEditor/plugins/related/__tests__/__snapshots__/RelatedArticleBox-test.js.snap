// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it goes in and out of edit mode 1`] = `
<div>
  <div
    aria-hidden="true"
    class="c-overlay"
  />
  <div
    style="height: 120px;"
  />
</div>
`;

exports[`it goes in and out of edit mode 2`] = `
<div
  class="css-mlcrrv"
  data-testid="relatedWrapper"
  role="button"
  tabindex="0"
>
  <section
    class="c-related-articles"
  >
    <h1
      class="c-section-heading c-related-articles__component-title"
    >
      injected
    </h1>
    <div
      class="c-related-articles__articles"
    >
      <article
        class="c-related-articles__item c-related-articles__item--external-learning-resources"
      >
        <h1
          class="c-related-articles__title"
        >
          <div
            class="c-content-type-badge c-content-type-badge--external-learning-resources c-content-type-badge--small c-content-type-badge--background"
          >
            <svg
              class="c-icon"
              data-license="Apache License 2.0"
              data-source="Material Design"
              fill="currentColor"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style="vertical-align: middle;"
              viewBox="0 0 24 24"
              width="1em"
            >
              <title>
                ExternalLearningResource
              </title>
              <g>
                <title>
                  Eksterne ressurser
                </title>
                <g
                  fill="none"
                  fill-rule="evenodd"
                  id="Page-1"
                  stroke="none"
                  stroke-width="1"
                >
                  <g
                    id="delte-ressurser"
                  >
                    <g
                      id="Shape"
                    >
                      <polygon
                        points="0 0 24 0 24 24 0 24"
                      />
                      <path
                        d="M3.9,12 C3.9,10.29 5.29,8.9 7,8.9 L11,8.9 L11,7 L7,7 C4.24,7 2,9.24 2,12 C2,14.76 4.24,17 7,17 L11,17 L11,15.1 L7,15.1 C5.29,15.1 3.9,13.71 3.9,12 Z M8,13 L16,13 L16,11 L8,11 L8,13 Z M17,7 L13,7 L13,8.9 L17,8.9 C18.71,8.9 20.1,10.29 20.1,12 C20.1,13.71 18.71,15.1 17,15.1 L13,15.1 L13,17 L17,17 C19.76,17 22,14.76 22,12 C22,9.24 19.76,7 17,7 Z"
                        fill="currentColor"
                        fill-rule="nonzero"
                      />
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <span
            class="c-related-articles__link-wrapper"
          >
            <a
              class="c-related-articles__link"
              href="https://www.vg.no"
              rel="noopener noreferrer"
              target="_blank"
            />
          </span>
        </h1>
        <p
          class="c-related-articles__description"
        />
        <p
          class="c-related-articles__link-info"
        >
          Nettside hos www.vg.no
        </p>
      </article>
    </div>
  </section>
</div>
`;
